### PR TITLE
rules: Action: Add a virtual method dump_json

### PR DESCRIPTION
**Pull request by fnareoh**
Add an optional virtual method to output the json dump for an action, in
order to make the dump process more elegant for the rules that use this feature.

This way for the list of actions to dump you can just call `action.dump_json(gamestate)`.

### CODE REVIEW

File: tools/generator/templates/rules/src/action_template.hh.jinja2 [hunk 1]
```diff
@@ -23,6 +23,7 @@ public:
 
     int check(const GameState& st) const override;
     void apply_on(GameState* st) const override;
+    void dump_json(const GameState& st, std::ostream& ss) const;
```
Comment (multun) [inline tools/generator/templates/rules/src/action_template.hh.jinja2:26]: I think I'd be slightly better with an `override`

**Review by multun — APPROVED** [review]
LGTM

**Comment by seirl** [comment]
Hmm, why don't we reuse handle_buffer instead of having to write that function manually?

**Review by seirl — CHANGES_REQUESTED** [review]
As explained above, this should use handle_buffer instead of reimplementing twice the serialization of each action.

**Comment by juli0z** [comment]
Migrated to https://gitlab.com/prologin/tech/tools/stechec2/-/merge_requests/138

**Pull request closed without merging**